### DISCUSSION TITLE
关于 Function Calls 输出不符合 OpenApi 规范的问题

### DISCUSSION THREAD

**Discussion by MapleEve** [zai-org/ChatGLM3#990] [Bad Case]
> ### 图片base64需要移除前缀；
> 智谱不支持 `data:image/png;base64,UklGRqTNAABXRUJQVlA4WAoAAAA` ，需要去掉前面的 `data:image/png;base64,`
> 
> ### `tool_calls` 流式有问题：
> 标准 OpenAI 的 tool_calls 输出如下：
> 
> ```json
> chunk 0
> {"choices":[{"index":0,"delta":{"role":"assistant","content":null,"tool_calls":[{"index":0,"id":"call_EOjIHZ8INcmtndup397Iy2R0","type":"function","function":{"name":"realtime-weather____fetchCurrentWeather","arguments":""}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 1
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"{\""}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 2
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"city"}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 3
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"\":\""}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 4
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"Hang"}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 5
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"zhou"}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 6
> {"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"\"}"}}]},"logprobs":null,"finish_reason":null}]}
> 
> chunk 7
> {"choices":[{"index":0,"delta":{},"logprobs":null,"finish_reason":"tool_calls"}]}
> ```
> 
> 但 zhipu 的接口输出格式如下：
> 
> ```json
> chunk 0
> {"choices":[{"index":0,"finish_reason":"tool_calls","delta":{"role":"assistant","tool_calls":[{"id":"call_8349638559205414584","index":0,"type":"function","function":{"name":"realtime-weather____fetchCurrentWeather","arguments":"{\"city\":\"杭州\"}"}}]}}]}
> 
> chunk 1
> {"choices":[{"index":0,"finish_reason":"tool_calls","delta":{"role":"assistant","content":""}}],"usage":{"prompt_tokens":153,"completion_tokens":20,"total_tokens":173}}
> ```
> 
> 可以看到最后的一个输出的 chunk 是，delta 是 `{"role":"assistant","content":""}`，这个行为和 openai 的不一致。这就导致最终前端收到的 arguments 参数为空字符串，所以会出现问题

**Comment by huangdehong** [comment]
关注一下

**Comment by zRzRzRzRzRzRzR** [comment]
这个是不兼容的，zhipuAI 这个库和OpenAI确实有区别，API部门还在想办法

**Comment by MapleEve** [comment]
> 这个是不兼容的，zhipuAI 这个库和OpenAI确实有区别，API部门还在想办法

确实还是尽量考虑同质化吧，其它大厂也都是偏向于同质化的API，这样接入、迁移都没有什么成本
